hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/ipc/Client.java

```diff
@@ -1153,9 +1087,51 @@ public void run() {
             + connections.size());
     }
 
+    /**
+     * A thread to write rpc requests to the socket.
+     */
+    private class RpcRequestSender implements Runnable {
+      @Override
+      public void run() {
+        while (!shouldCloseConnection.get()) {
+          ResponseBuffer buf = null;
+          try {
+            Pair<Call, ResponseBuffer> pair = rpcRequestQueue.take();
```

Comment:
There is a critical bug after we released this feature in our prod.

`take()` caused `RpcRequestSender` thread leaked if the connection is normally idled by itself.

Maybe we should change `take()` to `poll()` as bellows:
```
Pair<Call, ResponseBuffer> pair = rpcRequestQueue.poll(10, TimeUnit.MILLISECONDS);
if (pair == null) {
   continue;
}
```

Reply:
@ZanderXu Thanks for pointing this. i was about to backport it. Were you able to fix it?

Reply:
@mnpoonia We have cherrypick this PR and fix bugs, and deploy it to our prod environment. 

@omalley Sir, do you have time to push it forward? and can I fix it and push a new PR?

Reply:
I resolved the problem as you said by moving to a poll.